sdk/go/pulumi/provider.go

```diff
@@ -202,21 +224,316 @@ func constructInputsMap(ctx *Context, inputs map[string]interface{}) (Map, error
 			resultType = ot.(reflect.Type)
 		}
 
-		output := ctx.newOutput(resultType, ci.deps...)
+		output := ctx.newOutput(resultType, ci.Dependencies(ctx)...)
 		output.getState().resolve(value, known, secret, nil)
 		result[k] = output
 	}
 	return result, nil
 }
 
+func gatherDeps(v resource.PropertyValue, deps urnSet) {
+	switch {
+	case v.IsSecret():
+		gatherDeps(v.SecretValue().Element, deps)
+	case v.IsComputed():
+		gatherDeps(v.Input().Element, deps)
+	case v.IsOutput():
+		for _, urn := range v.OutputValue().Dependencies {
+			deps.add(URN(urn))
+		}
+		gatherDeps(v.OutputValue().Element, deps)
+	case v.IsResourceReference():
+		deps.add(URN(v.ResourceReferenceValue().URN))
+	case v.IsArray():
+		for _, e := range v.ArrayValue() {
+			gatherDeps(e, deps)
+		}
+	case v.IsObject():
+		for _, e := range v.ObjectValue() {
+			gatherDeps(e, deps)
+		}
+	}
+}
+
+func copyInputTo(ctx *Context, v resource.PropertyValue, dest reflect.Value) error {
+	contract.Assert(dest.CanSet())
+
+	// Check for nils. The destination will be left with the zero value.
+	if v.IsNull() {
+		return nil
+	}
+
+	switch {
+	case v.IsSecret():
+		known, element := true, v.SecretValue().Element
+		// See if it's value is unknown.
+		if element.IsComputed() {
+			known, element = false, element.Input().Element
+		}
+		// Handle this as a secret output.
+		return copyInputTo(ctx, resource.NewOutputProperty(resource.Output{
+			Element: element,
+			Known:   known,
+			Secret:  true,
+		}), dest)
+	case v.IsComputed():
+		// Handle this as an unknown output.
+		return copyInputTo(ctx, resource.MakeOutput(v.Input().Element), dest)
+	case v.IsOutput():
+		// If it's known, there aren't any dependencies, and it's not a secret, just copy the value.
+		if v.OutputValue().Known && len(v.OutputValue().Dependencies) == 0 && !v.OutputValue().Secret {
+			return copyInputTo(ctx, v.OutputValue().Element, dest)
+		}
+
+		if !dest.Type().Implements(outputType) && !dest.Type().Implements(inputType) {
+			return fmt.Errorf("expected destination type to implement %v or %v, got %v", inputType, outputType, dest.Type())
+		}
+
+		resourceDeps := make([]Resource, len(v.OutputValue().Dependencies))
+		for i, dep := range v.OutputValue().Dependencies {
+			resourceDeps[i] = ctx.newDependencyResource(URN(dep))
+		}
+
+		result, err := createOutput(ctx, dest.Type(), v.OutputValue().Element, v.OutputValue().Known,
+			v.OutputValue().Secret, resourceDeps)
+		if err != nil {
+			return err
+		}
+		dest.Set(result)
+		return nil
+	}
+
+	if dest.Type().Implements(outputType) {
+		result, err := createOutput(ctx, dest.Type(), v, true /*known*/, false /*secret*/, nil)
+		if err != nil {
+			return err
+		}
+		dest.Set(result)
+		return nil
+	}
+
+	if dest.Type().Implements(inputType) {
+		// Try to determine the input type from the interface.
+		if it, ok := inputInterfaceTypeToConcreteType.Load(dest.Type()); ok {
+			inputType := it.(reflect.Type)
```

Comment:
Rambling here, this is getting interesting right here:

Yay so we're deserializing PropertyValue into dest placeholder of  type `dest.Type() = Input[T]`, which is hairy because `Input` is an  interface.

No matter, homegrown typeclasses to the rescue (`inputInterfaceTypeToConcreteType`)!

Now we found the default type `inputType = D` that implements `D <: Input[T]`. We know that a value of type `D` can be put into a placeholder of type `Input[T]`.

So now we will just deserialize `v` of type `resource.PropertyValue` into a placeholder of type `D`.

QQ: do we need brand new code to do this last bit? I thought the codebase already knows how to do this? Or is it non-trivial to "view" a placeholder of type `Input[T]` as a placeholder of type `D`?

Reply:
> do we need brand new code to do this last bit?

I suspect we can indeed unify/simplify the code here, and I'll be taking a look at that subsequently.